pressKey('tab') sometimes works wrong with tabIndex=-1

### What is your Test Scenario?
There is a dialog popup and container with `tabIndex=-1` in it. There are some inputs in the container.
 
### What is the Current behavior?
When the Tab key is pressed, a focus goes outside the container.
 
### What is the Expected behavior?
When the Tab key is pressed, a focus should keep inside the container, just like it does without TestCafe.
 
### What is your web application and your TestCafe test code?
https://github.com/alexey-lin/testcafe-studio/commit/7b2cb2e0813506fc109acbda6c66285fee95aabc

[Doubling of submit\\cancel events in dialogs - Regression - GH-391](https://github.com/alexey-lin/testcafe-studio/blob/7b2cb2e0813506fc109acbda6c66285fee95aabc/test/functional/fixtures/dialogs/common.js#L51) test

<details>
<summary>Your complete test report:</summary>

<!-- Paste your complete result test report here (even if it is huge): -->
```
 Common tests
 × Doubling of submit\cancel events in dialogs - Regression - GH-391

   1) AssertionError: expected true to be falsy

  	Browser: Electron 7.1.9 / Windows 10

     	59 |    	.expect(await
  	isDialogNotVisible(dialogs.renameFile)).notOk()
     	60 |    	.click(Selector('.tc-file-rename-dialog'), { offsetX:
  	1, offsetY: 1 })
     	61 |    	.pressKey('tab esc');
     	62 |
     	63 |	await t
   	> 64 |    	.expect(await
  	isDialogNotVisible(dialogs.renameFile)).notOk()
     	65 |    	.click(dialogs.renameFile.closeBtn);
     	66 |
     	67 |	await deleteFixture(fileName);
     	68 |});
     	69 |

     	at <anonymous>
  	(C:\work\testcafe-studio\test\functional\fixtures\dialogs\common.js:64:63)



 1/1 failed (25s)
```
</details>

<details>
<summary>Screenshots:</summary>
<!-- If applicable, add screenshots to help explain the issue. -->
![2020-03-06_1414](https://user-images.githubusercontent.com/13483856/76079256-5485de00-5fb5-11ea-8bbf-a9ea285519a9.png)
</details>

### Steps to Reproduce:
<!-- Describe what we should do to reproduce the behavior you encountered. -->

1. Run the test.
3. Launch the app and repeat the same steps manually.
4. See the difference.
 
### Your Environment details:
 
* testcafe version:                   1.8.3-alpha.2
* node.js version:                    v12.14.1
* command-line arguments: gulp test-functional -s
* browser name and version: Electron 7.1.9
* platform and version:          Windows 10

--- a/src/client/automation/playback/press/utils.js
+++ b/src/client/automation/playback/press/utils.js
@@ -153,8 +153,12 @@ export function getNextFocusableElement (element, reverse, skipRadioGroups) {
     const currentIndex         = arrayUtils.indexOf(allFocusable, element);
     const isLastElementFocused = reverse ? currentIndex === 0 : currentIndex === allFocusable.length - 1;
 
-    if (isLastElementFocused)
+    if (isLastElementFocused) {
+        if (!reverse && element.tabIndex < 0)
+            return allFocusable.find(el => el.tabIndex === 0);
+
         return skipRadioGroups || !isRadioInput ? document.body : allFocusable[allFocusable.length - 1 - currentIndex];
+    }
 
     if (reverse && currentIndex === -1)
         return allFocusable[allFocusable.length - 1];

--- a/src/client/automation/playback/press/utils.js
+++ b/src/client/automation/playback/press/utils.js
@@ -155,7 +155,7 @@ export function getNextFocusableElement (element, reverse, skipRadioGroups) {
 
     if (isLastElementFocused) {
         if (!reverse && element.tabIndex < 0)
-            return allFocusable.find(el => el.tabIndex === 0);
+            return arrayUtils.find(allFocusable, el => el.tabIndex === 0);
 
         return skipRadioGroups || !isRadioInput ? document.body : allFocusable[allFocusable.length - 1 - currentIndex];
     }

--- a/src/client/core/utils/dom.js
+++ b/src/client/core/utils/dom.js
@@ -159,16 +159,38 @@ function sortBy (property) {
     };
 }
 
+function canFocus (element, activeElement, tabIndex) {
+    if (element === activeElement)
+        return true;
+
+    if (element.disabled)
+        return false;
+
+    if (getElementStyleProperty(element, 'display') === 'none' || getElementStyleProperty(element, 'visibility') === 'hidden')
+        return false;
+
+    if ((browserUtils.isIE || browserUtils.isAndroid) && isOptionElement(element))
+        return false;
+
+    if (tabIndex !== null && tabIndex < 0)
+        return false;
+
+    return true;
+}
+
 export function getFocusableElements (doc, sort = false) {
     // NOTE: We don't take into account the case of embedded contentEditable
     // elements and specify the contentEditable attribute for focusable elements
-    const allElements         = doc.querySelectorAll('*');
-    const invisibleElements   = getInvisibleElements(allElements);
-    const inputElementsRegExp = /^(input|button|select|textarea)$/;
-    const focusableElements   = [];
-    let element               = null;
-    let tagName               = null;
-    let tabIndex              = null;
+    const allElements           = doc.querySelectorAll('*');
+    const activeElement         = nativeMethods.documentActiveElementGetter.call(doc);
+    const activeElementTabIndex = getTabIndexAttributeIntValue(activeElement);
+    const invisibleElements     = getInvisibleElements(allElements);
+    const inputElementsRegExp   = /^(input|button|select|textarea)$/;
+    const focusableElements     = [];
+
+    let element  = null;
+    let tagName  = null;
+    let tabIndex = null;
 
     let needPush = false;
 
@@ -178,16 +200,7 @@ export function getFocusableElements (doc, sort = false) {
         tabIndex = getTabIndexAttributeIntValue(element);
         needPush = false;
 
-        if (element.disabled)
-            continue;
-
-        if (getElementStyleProperty(element, 'display') === 'none' || getElementStyleProperty(element, 'visibility') === 'hidden')
-            continue;
-
-        if ((browserUtils.isIE || browserUtils.isAndroid) && isOptionElement(element))
-            continue;
-
-        if (tabIndex !== null && tabIndex < 0)
+        if (!canFocus(element, activeElement, tabIndex))
             continue;
 
         if (inputElementsRegExp.test(tagName))
@@ -211,6 +224,9 @@ export function getFocusableElements (doc, sort = false) {
 
     //NOTE: remove children of invisible elements
     let result = arrayUtils.filter(focusableElements, el => !containsElement(invisibleElements, el));
+
+    if (activeElementTabIndex && activeElementTabIndex < 0)
+        sort = false;
 
     if (sort)
         result = sortElementsByFocusingIndex(result);

--- a/test/functional/fixtures/regression/gh-4848/pages/index.html
+++ b/test/functional/fixtures/regression/gh-4848/pages/index.html
@@ -1,0 +1,32 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <meta charset="UTF-8">
+    <title>gh-4848</title>
+</head>
+<body>
+    <button id="ft5" tabindex="-1">Focus "target 5"</button>
+    <button id="ft1" tabindex="-1">Focus "target 1"</button>
+    <div id="target1" tabindex="0">target 1</div>
+    <div id="target2" tabindex="-1">target 2</div>
+    <div id="target3">target 3</div>
+    <a href="http://example.com">test</a>
+    <div id="target4" tabindex="2">target 4</div>
+    <div id="target5" tabindex="-1">target 5</div>
+
+    <script>
+        document.querySelector('#target2').focus();
+
+        document.querySelector('#ft5').addEventListener('click', function () {
+            document.querySelector('#target5').focus();
+        });
+
+        document.querySelector('#ft1').addEventListener('click', function () {
+            var t1 = document.querySelector('#target1');
+
+            t1.setAttribute('tabindex', '-1');
+            t1.focus();
+        });
+    </script>
+</body>
+</html>

--- a/test/functional/fixtures/regression/gh-4848/test.js
+++ b/test/functional/fixtures/regression/gh-4848/test.js
@@ -1,0 +1,25 @@
+describe('[Regression](GH-4848) - Should focus next element if current element has negative tabIndex', function () {
+    it('Straight order. Middle', function () {
+        return runTests('testcafe-fixtures/index.js', 'Straight order. Middle');
+    });
+
+    it('Reversed order. Middle', function () {
+        return runTests('testcafe-fixtures/index.js', 'Reversed order. Middle');
+    });
+
+    it('Reversed order. First', function () {
+        return runTests('testcafe-fixtures/index.js', 'Reversed order. First');
+    });
+
+    it('Reversed order. Last', function () {
+        return runTests('testcafe-fixtures/index.js', 'Reversed order. Last');
+    });
+
+    it('Straight order. First', function () {
+        return runTests('testcafe-fixtures/index.js', 'Straight order. First');
+    });
+
+    it('Straight order. Last', function () {
+        return runTests('testcafe-fixtures/index.js', 'Straight order. Last');
+    });
+});

--- a/test/functional/fixtures/regression/gh-4848/testcafe-fixtures/index.js
+++ b/test/functional/fixtures/regression/gh-4848/testcafe-fixtures/index.js
@@ -1,0 +1,48 @@
+import { Selector } from 'testcafe';
+
+fixture `[Regression](GH-4848) - Should focus next element if current element has negative tabIndex`
+    .page `http://localhost:3000/fixtures/regression/gh-4848/pages/index.html`;
+
+const body = Selector('body');
+const ft1  = Selector('#ft1');
+const ft5  = Selector('#ft5');
+const a    = Selector('a');
+
+test(`Straight order. Middle`, async t => {
+    await t.pressKey('tab');
+    await t.expect(Selector(a).focused).eql(true);
+});
+
+test(`Reversed order. Middle`, async t => {
+    await t.pressKey('shift+tab');
+    await t.expect(Selector('#target1').focused).eql(true);
+});
+
+test(`Reversed order. Last`, async t => {
+    await t.click(ft5);
+    await t.pressKey('shift+tab');
+    await t.expect(Selector('#target4').focused).eql(true);
+});
+
+test(`Straight order. First`, async t => {
+    await t.click(ft1);
+    await t.pressKey('tab');
+    await t.expect(a.focused).eql(true);
+});
+
+test(`Reversed order. First`, async t => {
+    await t.click(ft1);
+    await t.pressKey('shift+tab');
+    await t.expect(body.focused).eql(true);
+});
+
+test(`Straight order. Last`, async t => {
+    await t.click(ft5);
+    await t.pressKey('tab');
+    await t.expect(Selector('#target1').focused).eql(true);
+
+    await t.click(ft1); // NOTE: make tabIndex of `target1` to -1
+    await t.click(ft5);
+    await t.pressKey('tab');
+    await t.expect(Selector(a).focused).eql(true);
+});

